synapse/storage/schema/main/delta/74/03thread_notifications_not_null.sql.postgres

```diff
@@ -0,0 +1,36 @@
+/* Copyright 2022 The Matrix.org Foundation C.I.C
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- The thread_id columns can now be made non-nullable, this is done by using a
+-- constraint (and not altering the column) to avoid taking out a full table lock.
+--
+-- We initially add an invalid constraint which guards against new data (this
+-- doesn't lock the table).
+ALTER TABLE event_push_actions_staging
+    ADD CONSTRAINT event_push_actions_staging_thread_id CHECK (thread_id IS NOT NULL) NOT VALID;
+ALTER TABLE event_push_actions
+    ADD CONSTRAINT event_push_actions_thread_id CHECK (thread_id IS NOT NULL) NOT VALID;
+ALTER TABLE event_push_summary
+    ADD CONSTRAINT event_push_summary_thread_id CHECK (thread_id IS NOT NULL) NOT VALID;
+
+-- We then validate the constraint which doesn't need to worry about new data and
+-- only needs a SHARE UPDATE EXCLUSIVE lock.
+ALTER TABLE event_push_actions_staging VALIDATE CONSTRAINT event_push_actions_staging_thread_id;
+ALTER TABLE event_push_actions VALIDATE CONSTRAINT event_push_actions_thread_id;
+ALTER TABLE event_push_summary VALIDATE CONSTRAINT event_push_summary_thread_id;
```

Comment:
Meaning that you want to ensure that a restart doesn't block attempting to validate these constraints? I'm fairly certain there would be no issue with doing that in the background.

Reply:
Yeah, exactly. Though I can imagine that its super quick due to the indices making putting them in a BG update unnecessary? But I'd like for us to either a) validate its quick, or b) move it to a BG update.

Reply:
Not quite the same, but I did some tests on a large-ish database I have for `event_push_actions` with 340,977 rows in it.

* Adding the `NOT VALID` constraint took 15.285 ms
* Validating the constraint took 90.614 ms

Unfortunately you can't `EXPLAIN` an `ALTER TABLE` so I don't know how exactly this works.

Note that matrix.org currently has 63,836,647 so that is almost 200x larger.

Reply:
We tested this on matrix.org and it took too long; we're going to move the validation constraint into a background job.